wcfsetup/install/files/lib/system/event/listener/AbstractEventListener.class.php

```diff
@@ -0,0 +1,48 @@
+<?php
+namespace wcf\system\event\listener;
+use wcf\data\AbstractDatabaseObjectAction;
+
+/**
+ * Preferred implementation for event listeners that dynamically invokes methods based on a predictable
+ * naming schema derived from the event name. In addition, `AbstractDatabaseObjectAction` supports deep
+ * method invocations for the generic action events. However, when there is no specific method, it will
+ * attempt to invoke the regular method instead.
+ * 
+ * Examples:
+ * Regular classes
+ *   eventName: makeSnafucated
+ *   derived method: onMakeSnafucated()
+ * 
+ * Classes deriving from AbstractDatabaseObjectAction
+ *   eventName: initializeAction
+ *   actionName: makeSnafucated
+ *   derived method: onInitializeActionMakeSnafucated()
+ *   ATTENTION: If this method does not exist, it will attempt to invoke `onMakeSnafucated()` instead. 
+ *
+ * @author      Olaf Braun, Alexander Ebert
+ * @copyright   2001-2020 WoltLab GmbH
+ * @license     GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package     WoltLabSuite\Core\System\Event\Listener
+ */
+abstract class AbstractEventListener implements IParameterizedEventListener {
+	/**
+	 * @inheritDoc
+	 */
+	public final function execute($eventObj, $className, $eventName, array &$parameters) {
+		static $genericDboActionNames = ['finalizeAction', 'initializeAction', 'validateAction'];
```

Comment:
I would make this a class constant.

Reply:
I thought about this, but stepped back from it. The current approach encapsulates the entire logic inside this method without exposing a (public, PHP 7.0) constant. This also keeps the class as clean as possible without interfering with derived classes, making it an easier drop-in replacement for existing code.

Reply:
I don't particularly care either way.